File: curry/Views/TKInputView.m
//
//  TKInputView.m
//  Created by Devin Ross on 3/21/14.
//
/*
 
 curry || https://github.com/devinross/curry
 
 Permission is hereby granted, free of charge, to any person
 obtaining a copy of this software and associated documentation
 files (the "Software"), to deal in the Software without
 restriction, including without limitation the rights to use,
 copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the
 Software is furnished to do so, subject to the following
 conditions:
 
 The above copyright notice and this permission notice shall be
 included in all copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 OTHER DEALINGS IN THE SOFTWARE.
 
 */


#import "TKInputView.h"
#import "UIDevice+TKCategory.h"
#import "UIColor+TKCategory.h"
#import "TKInputKey.h"
#import "UIView+TKCategory.h"
#import "UIImageView+TKCategory.h"
#import "UIImage+TKCategory.h"
#import "TKFoundation.h"

@interface TKInputView ()

@property (nonatomic,strong) NSArray *keyViews;
@property (nonatomic,assign) TKInputKey *originalSelectedKey;
@property (nonatomic,strong) TKInputKey *selectedKey;
@property (nonatomic,strong) NSTimer *holdTimer;
@property (nonatomic,assign) NSInteger holdCounter;

@end

@implementation TKInputView



+ (CGSize) sizeOfKeyboardForMainScreen{
	CGFloat height = 352;
	if([UIDevice currentDevice].phoneIdiom){
		height = [UIScreen mainScreen].traitCollection.verticalSizeClass == UIUserInterfaceSizeClassCompact ? 162 :  216;
	}
	return CGSizeMake([UIScreen mainScreen].bounds.size.width, height);
}

- (instancetype) initWithFrame:(CGRect)frame withKeysModels:(NSArray*)keys{
	frame.size = [TKInputView sizeOfKeyboardForMainScreen];
	if(!(self=[super initWithFrame:frame])) return nil;
	
	self.keyViews = keys;
	self.autoresizingMask = UIViewAutoresizingFlexibleHeight | UIViewAutoresizingFlexibleWidth;
	self.backgroundColor = [UIColor colorWithHex:0xd7dadf];
	self.clipsToBounds = YES;
	self.multipleTouchEnabled = NO;
	self.exclusiveTouch = YES;
	
	CGRect cntFrame = frame;
	if([UIDevice currentDevice].padIdiom){
		cntFrame.size.width = 544;
		cntFrame.origin.x = round((self.frame.size.width - cntFrame.size.width) / 2);
	}else{
		cntFrame.origin = CGPointMake(0, 1);
		cntFrame.size.height -= 1;
	}
	
	
	
	self.containerView = [[UIView alloc] initWithFrame:cntFrame];
	self.containerView.clipsToBounds = YES;
	
	if([UIDevice currentDevice].padIdiom)
		self.containerView.autoresizingMask = UIViewAutoresizingFlexibleLeftMargin | UIViewAutoresizingFlexibleHeight | UIViewAutoresizingFlexibleRightMargin;
	else
		self.containerView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
	self.containerView.clipsToBounds = YES;
	[self addSubview:self.containerView];
	
	
	NSInteger tag = 0;
	
	for(TKInputKey *key in keys){
		key.tag = tag;
		[key setHighlighted:NO];
		[self.containerView addSubviewToBack:key];
		tag++;
	}
	
	
	
	
	if([UIDevice currentDevice].padIdiom){
		
		

        CGRect rect = CGRectMake(frame.size.width - 80 - 32, frame.size.height - 75 - 12, 80, 75);
		
		UIImage *img = [UIImage imageNamed:@"keyboard-down-keyboard" inBundle:[NSBundle bundleForClass:[self class]] compatibleWithTraitCollection:[UITraitCollection traitCollectionWithDisplayScale:[UIScreen mainScreen].scale]];

		self.hideKeyboardKey = [[TKInputKey alloc] initWithFrame:rect symbol:img normalType:TKInputKeyTypeDefault selectedType:TKInputKeyTypeHighlighted runner:NO];
		self.hideKeyboardKey.autoresizingMask = UIViewAutoresizingFlexibleLeftMargin | UIViewAutoresizingFlexibleTopMargin;
		self.hideKeyboardKey.tag = tag;
		[self.hideKeyboardKey setHighlighted:NO];
		[self addSubview:self.hideKeyboardKey];
		
		UIImage *dotsImage = [UIImage imageNamed:@"keyboard-move-keyboard-dots" inBundle:[NSBundle bundleForClass:[self class]] compatibleWithTraitCollection:[UITraitCollection traitCollectionWithDisplayScale:[UIScreen mainScreen].scale]];
		UIImageView *dots = [UIImageView imageViewWithFrame:CGRectMakeWithSize(CGRectGetWidth(frame) - 18, CGRectGetHeight(frame) - 57, dotsImage.size)];
		dots.image = dotsImage;
		dots.autoresizingMask = UIViewAutoresizingFlexibleLeftMargin | UIViewAutoresizingFlexibleTopMargin;
		[self addSubview:dots];
	}
	
	[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(didHideNotification:) name:UIKeyboardDidHideNotification object:nil];
	
    return self;
}

- (void) dealloc{
	[[NSNotificationCenter defaultCenter] removeObserver:self name:UIKeyboardDidHideNotification object:nil];
}

- (void) didHideNotification:(id)sender{

	if(self.traitCollection.userInterfaceIdiom == UIUserInterfaceIdiomPhone){
		CGFloat height = self.traitCollection.verticalSizeClass == UIUserInterfaceSizeClassCompact ? 162 :  216;
		CGRect rect = self.frame;
		rect.size.height = height;
		self.frame = rect;
	}
	
}

- (void) touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event{
	
	UITouch *touch = [touches anyObject];
	UIView *view = [self hitTest:[touch locationInView:self] withEvent:event];
	
	if(self.selectedKey!=nil){
		[self.selectedKey setHighlighted:NO];
		self.originalSelectedKey = nil;
		self.selectedKey = nil;
	}
	
	if(![view isKindOfClass:[TKInputKey class]]) return;
	[[UIDevice currentDevice] playInputClick];
	
	self.originalSelectedKey = (TKInputKey*)view;
	self.selectedKey = self.originalSelectedKey;
	[self.selectedKey setHighlighted:YES];
	[self.containerView bringSubviewToFront:self.selectedKey];
    
    if(self.originalSelectedKey.canTapAndHold){
        self.holdCounter = 0;
        [self.holdTimer invalidate];
        self.holdTimer = nil;
        self.holdTimer = [NSTimer scheduledTimerWithTimeInterval:0.35 target:self selector:@selector(holding:) userInfo:nil repeats:YES];
    }
	
}
- (void) touchesMoved:(NSSet *)touches withEvent:(UIEvent *)event{
    
	UITouch *touch = [touches anyObject];
	UIView *view = [self hitTest:[touch locationInView:self] withEvent:event];
    
	TKInputKey *currentKey = nil;
	if([view isKindOfClass:[TKInputKey class]])
		currentKey = (TKInputKey*)view;
	
	BOOL on = NO;
	if(self.originalSelectedKey == currentKey)
		on = YES;
	else if(self.originalSelectedKey.runner && currentKey.runner)
		on = YES;
	
	
	if(self.selectedKey != currentKey){
		[self.selectedKey setHighlighted:NO];
		self.selectedKey = nil;
	}
	
	if(on){
		self.selectedKey = currentKey;
		[self.containerView bringSubviewToFront:self.selectedKey];
		[self.selectedKey setHighlighted:YES];
	}
	
}

- (void) holding:(id)sender{
    
    if(self.originalSelectedKey == self.selectedKey){
        self.holdCounter++;
        if(self.holdCounter < 6){
            [self sendTapKeyEvent];
        }else{
            [self sendTapKeyEvent];
            [self sendTapKeyEvent];
            [self sendTapKeyEvent];
            [self sendTapKeyEvent];
            [self sendTapKeyEvent];
        }
    }
    
}

- (void) sendTapKeyEvent{
    
    if(self.selectedKey == self.backspaceKey){
        
        BOOL delete = YES;
        if(self.textField.text.length < 1)
            delete = NO;
        else if(self.textField.delegate && [self.textField.delegate respondsToSelector:@selector(textField:shouldChangeCharactersInRange:replacementString:)])
            delete = [self.textField.delegate textField:self.textField shouldChangeCharactersInRange:NSMakeRange(self.textField.text.length-1, 1) replacementString:@""];
        
        if(delete) [self.textField deleteBackward];
        
    }else if(self.selectedKey){
        [self.delegate inputView:self didSelectKey:self.selectedKey];
    }
}

- (void) touchesEnded:(NSSet *)touches withEvent:(UIEvent *)event{
	
    if(self.holdTimer == nil || self.holdCounter < 1)
        [self sendTapKeyEvent];

	[self touchesCancelled:touches withEvent:event];
}
- (void) touchesCancelled:(NSSet *)touches withEvent:(UIEvent *)event{
	
	[self.selectedKey setHighlighted:NO];
	self.selectedKey = nil;
	self.originalSelectedKey = nil;
    [self.holdTimer invalidate];
    self.holdTimer = nil;
    self.holdCounter = 0;
	
}
- (BOOL) enableInputClicksWhenVisible{
    return YES;
}

@end

File: CurrySwift/Views/TKInputView.swift
//
//  TKInputView.swift
//  Created by Devin Ross on 3/21/14.
//
/*

curry || https://github.com/devinross/curry

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

*/

import UIKit


extension Bundle {
	
	/// The bundle carrying curry's keyboard images.
	static var curryResources: Bundle {
		#if SWIFT_PACKAGE
		return .module
		#else
		return Bundle(for: TKInputView.self)
		#endif
	}
	
	/// An image from that bundle, at the main screen's scale.
	static func curryImage(named name: String) -> UIImage? {
		UIImage(named: name, in: curryResources,
		        compatibleWith: UITraitCollection(displayScale: UIScreen.main.scale))
	}
	
}


@objc(TKInputViewDelegate)
public protocol TKInputViewDelegate: NSObjectProtocol {
	@objc func inputView(_ inputView: TKInputView, didSelectKey key: TKInputKey)
}


/** A custom keyboard made of `TKInputKey` views. */
@objc(TKInputView)
open class TKInputView: UIView, UIInputViewAudioFeedback {
	
	@objc open weak var delegate: TKInputViewDelegate?
	@objc open weak var textField: UITextField?
	
	@objc open var backspaceKey: TKInputKey
	@objc open var hideKeyboardKey: TKInputKey?
	@objc public private(set) var selectedKey: TKInputKey?
	@objc open var containerView: UIView!
	
	private weak var originalSelectedKey: TKInputKey?
	private var holdTimer: Timer?
	private var holdCounter: Int = 0
	private var originalFrame: CGRect = .zero
	
	@objc open class var sizeOfKeyboardForMainScreen: CGSize {
		var height: CGFloat = 352
		if UIDevice.current.userInterfaceIdiom == .phone {
			height = UIScreen.main.traitCollection.verticalSizeClass == .compact ? 162 : 216
		}
		return CGSize(width: UIScreen.main.bounds.size.width, height: height)
	}
	
	public convenience init() {
		self.init(frame: CGRect(origin: .zero, size: TKInputView.sizeOfKeyboardForMainScreen), keysModels: [])
	}
	
	public convenience override init(frame: CGRect) {
		self.init(frame: frame, keysModels: [])
	}
	
	@objc(initWithFrame:withKeysModels:)
	public init(frame: CGRect, keysModels keys: [TKInputKey]) {
		
		var sized = frame
		sized.size = TKInputView.sizeOfKeyboardForMainScreen
		
		// Built before super.init because Swift requires stored properties to be set
		// first; the Objective-C original created it lazily off self.frame.
		backspaceKey = TKInputKey(frame: CGRect(x: 0, y: 0, width: sized.width * 0.25, height: sized.height),
		                          symbol: Bundle.curryImage(named: "keyboard-backspace-key"),
		                          normalType: .default, selectedType: .dark, runner: false)
		backspaceKey.canTapAndHold = true
		backspaceKey.accessibilityLabel = NSLocalizedString("Delete", comment: "")
		
		super.init(frame: sized)
		
		originalFrame = sized
		autoresizingMask = [.flexibleHeight, .flexibleWidth]
		backgroundColor = UIColor.color(hex: 0xd7dadf)
		clipsToBounds = true
		
		#if !os(tvOS)
		isMultipleTouchEnabled = false
		isExclusiveTouch = true
		#endif
		
		var containerFrame = sized
		if UIDevice.current.padIdiom {
			containerFrame.size.width = 544
			containerFrame.origin.x = ((sized.width - containerFrame.width) / 2).rounded()
		} else {
			containerFrame.origin = CGPoint(x: 0, y: 1)
		}
		
		let container = UIView(frame: containerFrame)
		container.clipsToBounds = true
		if UIDevice.current.padIdiom {
			container.autoresizingMask = [.flexibleLeftMargin, .flexibleHeight, .flexibleRightMargin]
		}
		containerView = container
		addSubview(container)
		
		for key in keys { install(key: key) }
		
		if UIDevice.current.padIdiom {
			
			let rect = CGRect(x: sized.width - 80 - 32, y: sized.height - 75 - 12, width: 80, height: 75)
			let hide = TKInputKey(frame: rect, symbol: Bundle.curryImage(named: "keyboard-down-keyboard"),
			                      normalType: .default, selectedType: .highlighted, runner: false)
			hide.autoresizingMask = [.flexibleLeftMargin, .flexibleTopMargin]
			hide.tag = container.subviews.count
			hide.setHighlighted(false)
			hideKeyboardKey = hide
			addSubview(hide)
			
			if let dotsImage = Bundle.curryImage(named: "keyboard-move-keyboard-dots") {
				let dots = UIImageView(frame: CGRect(origin: CGPoint(x: sized.width - 18, y: sized.height - 57),
				                                     size: dotsImage.size))
				dots.image = dotsImage
				dots.autoresizingMask = [.flexibleLeftMargin, .flexibleTopMargin]
				addSubview(dots)
			}
		}
		
		#if !os(tvOS)
		NotificationCenter.default.addObserver(self, selector: #selector(didHideNotification(_:)),
		                                       name: UIResponder.keyboardDidHideNotification, object: nil)
		#endif
	}
	
	public required init?(coder: NSCoder) {
		backspaceKey = TKInputKey(frame: .zero, symbol: nil, normalType: .default, selectedType: .dark, runner: false)
		super.init(coder: coder)
	}
	
	/// Adds a key to the container, numbered and unhighlighted.
	@objc(installKey:)
	open func install(key: TKInputKey) {
		key.tag = containerView.subviews.count
		key.setHighlighted(false)
		containerView.addSubviewToBack(key)
	}
	
	@objc private func didHideNotification(_ sender: Any?) {
		guard traitCollection.userInterfaceIdiom == .phone else { return }
		frame.size.height = traitCollection.verticalSizeClass == .compact ? 162 : 216
	}
	
	open override func layoutSubviews() {
		super.layoutSubviews()
		
		var bottom: CGFloat = 0
		
		if safeAreaInsets.bottom > 0 {
			frame.size.height += safeAreaInsets.bottom
			frame.origin.y -= safeAreaInsets.bottom
			bottom = safeAreaInsets.bottom
		}
		
		guard traitCollection.userInterfaceIdiom == .phone else { return }
		
		let compact = traitCollection.verticalSizeClass == .compact
		let x = (bounds.width - 450) / 2
		var frame = bounds.insetBy(dx: compact ? x : 0, dy: 0)
		frame.origin.y += 1
		frame.size.height += 1
		frame.size.height -= bottom
		containerView.frame = frame
	}
	
	// MARK: - Touches
	
	open override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
		
		guard let touch = touches.first else { return }
		let view = hitTest(touch.location(in: self), with: event)
		
		if selectedKey != nil {
			selectedKey?.setHighlighted(false)
			originalSelectedKey = nil
			selectedKey = nil
		}
		
		guard let key = view as? TKInputKey else { return }
		UIDevice.current.playInputClick()
		
		originalSelectedKey = key
		selectedKey = key
		key.setHighlighted(true)
		containerView.bringSubviewToFront(key)
		
		if key.canTapAndHold {
			holdCounter = 0
			holdTimer?.invalidate()
			holdTimer = Timer.scheduledTimer(timeInterval: 0.35, target: self,
			                                 selector: #selector(holding(_:)), userInfo: nil, repeats: true)
		}
	}
	
	open override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
		
		guard let touch = touches.first else { return }
		let currentKey = hitTest(touch.location(in: self), with: event) as? TKInputKey
		
		var on = false
		if originalSelectedKey === currentKey {
			on = true
		} else if originalSelectedKey?.runner == true && currentKey?.runner == true {
			on = true
		}
		
		if selectedKey !== currentKey {
			selectedKey?.setHighlighted(false)
			selectedKey = nil
		}
		
		if on, let currentKey {
			selectedKey = currentKey
			containerView.bringSubviewToFront(currentKey)
			currentKey.setHighlighted(true)
		}
	}
	
	@objc private func holding(_ sender: Any?) {
		
		guard originalSelectedKey === selectedKey else { return }
		
		holdCounter += 1
		let repeats = holdCounter < 6 ? 1 : 5
		for _ in 0 ..< repeats { sendTapKeyEvent() }
	}
	
	/// Sends the selected key to the text field, respecting its delegate's veto.
	open func sendTapKeyEvent() {
		
		guard let selectedKey else { return }
		
		if selectedKey === backspaceKey {
			
			guard let textField, let text = textField.text, !text.isEmpty else { return }
			
			let allowed = textField.delegate?.textField?(textField,
			                                             shouldChangeCharactersIn: NSRange(location: (text as NSString).length - 1, length: 1),
			                                             replacementString: "") ?? true
			if allowed { textField.deleteBackward() }
			
		} else {
			delegate?.inputView(self, didSelectKey: selectedKey)
		}
	}
	
	open override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
		if holdTimer == nil || holdCounter < 1 {
			sendTapKeyEvent()
		}
		touchesCancelled(touches, with: event)
	}
	
	open override func touchesCancelled(_ touches: Set<UITouch>, with event: UIEvent?) {
		selectedKey?.setHighlighted(false)
		selectedKey = nil
		originalSelectedKey = nil
		holdTimer?.invalidate()
		holdTimer = nil
		holdCounter = 0
	}
	
	open var enableInputClicksWhenVisible: Bool { true }
	
	/// Lets subclasses adopt a key the touch handling already knows about.
	func select(key: TKInputKey?) {
		selectedKey = key
	}
	
}
